### PR TITLE
GW-1533: Add SourcesClient

### PR DESCRIPTION
**Pull request by philippe-leonhardt-cko**
[GW-1533](https://checkout.atlassian.net/browse/GW-1533)
Added implementation of a `SourcesClient` based on the existing `PaymentsClient` and `TokensClient`.

### CODE REVIEW

**Comment by philippe-leonhardt-cko** [comment]
Tests were failing, because merchant had not been onboarded yet. Rerunning tests should pass CI now.

File: src/CheckoutSdk/Sources/SourceResponse.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        /// <summary>
+        /// Gets or sets the response code of the source.
+        /// </summary>
+        [JsonProperty(PropertyName = "response_code")]
+        public string ResponseCode { get; set; }
+
+        /// <summary>
+        /// Gets or sets the customer of the source.
+        /// </summary>
+        public Dictionary<string, string> Customer { get; set; }
```
Comment (filip-soldan-cko) [inline src/CheckoutSdk/Sources/SourceResponse.cs]: Similarily to `SourceRequest`, please consider using `CustomerResponse` object

File: src/CheckoutSdk/Sources/SourceResponse.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+using Checkout.Common;
+using Newtonsoft.Json;
+using System.Collections.Generic;
+
+namespace Checkout.Sources
+{
+    /// <summary>
+    /// Defines a source response.
+    /// </summary>
+    public class SourceResponse
```
Comment (filip-soldan-cko) [inline src/CheckoutSdk/Sources/SourceResponse.cs]: This could inherit from `Resource` class in a similar manner as other response objects do.

File: src/CheckoutSdk/Sources/ISourcesClient.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Checkout.Sources
+{
+    /// <summary>
+    /// Defines the operations available on the Checkout.com Sources API.
+    /// </summary>
+    public interface ISourcesClient
+    {
+        /// <summary>
+        /// Request a source.
```
Comment (filip-soldan-cko) [inline src/CheckoutSdk/Sources/ISourcesClient.cs]: Correct me if I'm wrong, but I understand you're implementing [this method](https://api-reference.checkout.com/#tag/Sources/paths/~1sources/post). If this is the case, what do you think about making the descriptions match?

**Comment by philippe-leonhardt-cko** [comment]
@filip-soldan-cko I have liaised internally about your suggestion to have `SourceResponse` mirror the 201/202 behavior of `PaymentResponse`.
If it is ok for you I would leave this for a future PR as we do not yet have any use case for async creation of a source.

File: src/CheckoutSdk/Sources/SourceRequest.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /// <summary>
+    /// Defines a request for a source.
+    /// </summary>
+    public class SourceRequest
+    {
+        /// <summary>
+        /// Creates a new default source request.
+        /// </summary>
+        public SourceRequest() { }
```
[the remark this answers is not in the document]
Comment (filip-soldan-cko) [reply src/CheckoutSdk/Sources/SourceRequest.cs]: Fair enough, seems like a valid use case. It won't work for `PaymentRequest` though.
We could still add the second constructor to indicate the requirement of those fields.

**Comment by filip-soldan-cko** [comment]
@philippe-leonhardt-cko I'd like to hold on to the specs as tightly as possible, especially because if we decide to go that route anytime in the future this could potentially introduce breaking changes for our SDK users. 
Perhaps we could introduce a stub implementation for the missing use case and raise a Github issue out of it so we don't forget about implementing it properly later?

**Comment by filip-soldan-cko** [comment]
@ben-foster-cko What's your take on this?

File: src/CheckoutSdk/Common/CustomerRequest.cs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿namespace Checkout.Payments
+﻿namespace Checkout.Common
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Common/CustomerRequest.cs]: This is a breaking change. I would recommend creating a specific customer request/response class specifically for sources or referencing the payments one for now until we can normalise these later.

File: src/CheckoutSdk/Common/CustomerResponse.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿namespace Checkout.Payments
+﻿namespace Checkout.Common
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Common/CustomerResponse.cs]: Breaking change

File: src/CheckoutSdk/Payments/AlternativePaymentSource.cs [hunk 7]
```diff
@@ -9,9 +9,15 @@ public class AlternativePaymentSource : Dictionary<string, string>, IRequestSour
     {
         private const string TypeField = "type";
 
+        /// <summary>
+        /// Creates a new default <see cref="AlternativePaymentSource"/> instance.
+        /// </summary>
+        public AlternativePaymentSource() { }
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Payments/AlternativePaymentSource.cs]: Why was this added? It should not be possible to create an `AlternativePaymentSource` without a `type`

File: src/CheckoutSdk/Sources/ResponseData.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+using Newtonsoft.Json;
+
+namespace Checkout.Sources
+{
+    /// <summary>
+    /// Defines the response data of a <see cref="SourceResponse"/>.
+    /// </summary>
+    public class ResponseData
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/ResponseData.cs]: Response data is dynamic and can vary between sources. This should be treated the same way as `AlternativePaymentSourceResponse`

File: src/CheckoutSdk/Sources/SourceData.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
+using Newtonsoft.Json;
+
+namespace Checkout.Sources
+{
+    /// <summary>
+    /// Defines the source data of a <see cref="SourceRequest"/>.
+    /// </summary>
+    public class SourceData
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/SourceData.cs]: Source data is dynamic and can vary between sources. This should be treated the same as `AlternativePaymentSource`

File: src/CheckoutSdk/Sources/SourcePending.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+using Checkout.Common;
+
+namespace Checkout.Sources
+{
+    /// <summary>
+    /// Indicates the source is pending, either for deferred processing or awaiting redirect.
+    /// </summary>
+    public class SourcePending : Resource
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/SourcePending.cs]: Please remove this as pending responses are not supported

File: src/CheckoutSdk/Sources/SourceProcessed.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        /// <summary>
+        /// Gets or sets the type of the source.
+        /// </summary>
+        public string Type { get; set; }
+
+        /// <summary>
+        /// Gets or sets the response code of the source.
+        /// </summary>
+        [JsonProperty(PropertyName = "response_code")]
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/SourceProcessed.cs]: These attributes are not necessary. The serializer already handles snake casing

File: src/CheckoutSdk/Sources/SourceProcessed.cs [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        /// <summary>
+        /// Gets or sets the <see cref="ResponseData"/> of the source.
+        /// </summary>
+        [JsonProperty(PropertyName = "response_data")]
+        public ResponseData ResponseData { get; set; }
+
+        /// <summary>
+        /// Gets or sets the links of the source.
+        /// </summary>
+        [JsonProperty(PropertyName = "_links")]
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/SourceProcessed.cs]: Don't remove this one though - it is necessary ;)

File: src/CheckoutSdk/Sources/SourceRequest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ben-foster-cko) [reply src/CheckoutSdk/Sources/SourceRequest.cs]: Whilst they could do this, they shouldn't nor should we allow it as it bypasses our validation.

File: src/CheckoutSdk/Sources/SourceResponse.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        /// <summary>
+        /// Gets or sets the processed response returned following a successfully processed source (HTTP Status Code 201).
+        /// </summary>
+        public SourceProcessed Source { get; set; }
+
+        /// <summary>
+        /// Gets or sets the pending response returned for asynchronous source or when further action such as a redirect is required (HTTP Status Code 202).
+        /// </summary>
+        public SourcePending Pending { get; set; }
```
Comment (ben-foster-cko) [inline src/CheckoutSdk/Sources/SourceResponse.cs]: Not supported though I think it's a good idea to leave this wrapper type in as it's possible we will have redirect sources in the future
